### ISSUE TITLE
[DNNL] Add inplace support to DNNL ops

### ISSUE THREAD

**Issue by jczaja** [PaddlePaddle/Paddle#22904]
DNNL primitives works faster if computation is in-place e.g. Input and output are the same memory
so input is overwritten by outcome of computation. We checked (using external program) that
DNNL's layer norm takes 30% less time to execute when in-place. We hope that other operators
will gain as well.

Following DNNL operators are supporting in place:
- Layer Norm
- Batch Norm
- softmax
- binary

Goal of this issue is to:
Add IR pass to scan a graph and  and make In = Out tensors, if input is not used by more than one operator (We cannot overwrite input of layer norm if this input is to be used in other operator as well)

**Comment by jczaja** [comment]
Here are  early results(latency in ms, lower is better) of experiments with softmax inplace mkl-dnn vs develop):
| Branch        | FP32, Threads: 1         | FP32, Threads: 20| Int8, Threads: 1 | Int8, Threads: 20|
| ------------- |:-------------:| -----:|--------:|-----:|
| develop      |  171.736 | 29.8331 | 64.6135  | 17.1636 |
| softmax dnnl inplace     | 170.047 | 29.6391 | 64.1744 | 16.7482 |

Platform: Intel Xeon 6248 , Turbo ON
Instructions how to was run: https://github.com/PaddlePaddle/benchmark/tree/master/Inference/c%2B%2B/ernie/mkldnn

**Comment by jczaja** [comment]
done
